pyproject.toml

```diff
@@ -24,6 +24,10 @@ exclude = [
 ]
 
 [tool.ruff.lint]
+# Disable all rules on test cases by default:
+# test cases often deliberately contain code
+# that might not be considered idiomatic or modern
+exclude = ["**/test_cases/**/*.py"]
```

Comment:
Are you sure about disabling flake8-future-annotations-fa, isort, probably eventually pycodestyles and possibly more eventually?

Reply:
- Definitely sure about disabling pycodestyle -- I think basically any antipatterns in test cases that black won't autofix are likely to be false positive that we're deliberately using in the test cases
- Agree that the FA codes are useful -- I added some stuff to `.pre-commit-config.yaml` to ensure that we still run those on the test cases, even though they're disabled in our `pyproject.toml` file by default

Reply:
I'll add a comment to `pyproject.toml` to make this clearer!

Reply:
I suppose isort might also be useful on these files -- that's a fair point.

Reply:
It feels a bit odd te me that local runs will differ and editors won't autofix (either from a manual call or on-save) or even visually indicate rule violations (yellow squiggle) that will be raised by CI.

As far as local runs go, maybe instead of running Ruff check, black, flake8, etc, the recommendation would be to run `pre-commit`, especially since the default command only runs on changed files. A lot of `scripts/runtests.py` could be replaced with a `pre-commit` call too.

But I'm still putoff a bit by having my editor stop showing me some issues. Granted it's only in test_cases, which already often differ in local runs with type-checkers anyway 😅

Reply:
> It feels a bit odd te me that local runs will differ and editors won't autofix (either from a manual call or on-save) or even visually indicate rule violations (yellow squiggle) that will be raised by CI.

I see your point... but the only rules we'll be enabling on the test cases are isort and flake8-annotations -- and they'll be autofixed in CI anyway? So I feel like it's not a _massive_ issue... typeshed is already kind-of a hopeless case when it comes to editors understanding our directory structure anyway, in my experience 😆

> That makes me think, as far as local runs go, maybe instead of running Ruff check, black, flake8, etc, the recommendation would be to run `pre-commit`, especially since the default command only runs on changed files. A lot of `scripts/runtests.py` could be replaced with a `pre-commit` call too.

Goood point -- I pushed some updates along those lines!

Reply:
At least now local runs will match the CI, which was a bigger concern for me. And yeah not a big deal if it only affects `test_cases`.

> typeshed is already kind-of a hopeless case when it comes to editors understanding our directory structure anyway, in my experience 😆

Honestly other than `test_cases`, stubs I migrated from https://github.com/microsoft/python-type-stubs, it hasn't been too bad. Some mypy/pyright discrepancies due to strictness level here and there.